server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -0,0 +1,226 @@
+            .features(Scalar.Feature.DETERMINISTIC)
+            .typeVariableConstraints(typeVariable("V"))
+            .build();
+
+    static {
+        DataTypes.register(TopKAggregation.TopKStateType.ID, in -> TopKAggregation.TopKStateType.INSTANCE);
+    }
+
+    public static void register(Functions.Builder builder) {
+        builder.add(
+            DEFAULT_SIGNATURE,
+            TopKAggregation::new
+        );
+
+        builder.add(
+            PARAMETER_SIGNATURE,
+            TopKAggregation::new
+        );
+    }
+
+    private final Signature signature;
+    private final BoundSignature boundSignature;
+
+    private TopKAggregation(Signature signature, BoundSignature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public BoundSignature boundSignature() {
+        return boundSignature;
+    }
+
+
+    @Nullable
+    @Override
+    public TopKState newState(RamAccounting ramAccounting,
+                              Version indexVersionCreated,
+                              Version minNodeInCluster,
+                              MemoryManager memoryManager) {
+        return new TopKState(new ItemsSketch<>(32), 8);
+    }
+
+    @Override
+    public TopKState iterate(RamAccounting ramAccounting,
+                             MemoryManager memoryManager,
+                             TopKState state,
+                             Input<?>... args) throws CircuitBreakingException {
+        if (state.itemsSketch.isEmpty() && args.length == 2) {
+            Integer limit = (Integer) args[1].value();
+            if (limit <= 0) {
+                throw new IllegalArgumentException("Invalid value for parameter k for top-k aggregation");
+            }
+            state = new TopKState(new ItemsSketch<>(maxMapSize(limit)), limit);
+        }
+        Object value = args[0].value();
+        state.itemsSketch.update(value);
+        return state;
+    }
+
+    private int maxMapSize(int x) {
+        // max map size should be 4 * the limit based on the power of 2 to avoid errors
+        return (int) Math.pow(2, Math.ceil(Math.log(x) / Math.log(2))) * 4;
+    }
+
+    @Override
+    public TopKState reduce(RamAccounting ramAccounting, TopKState state1, TopKState state2) {
+        return new TopKState(state1.itemsSketch.merge(state2.itemsSketch), Math.min(state1.limit, state2.limit));
+    }
+
+    @Override
+    public List<Object> terminatePartial(RamAccounting ramAccounting, TopKState state) {
+        ItemsSketch.Row<Object>[] frequentItems = state.itemsSketch.getFrequentItems(ErrorType.NO_FALSE_NEGATIVES);
+        int limit = Math.min(frequentItems.length, state.limit);
+        var result = new ArrayList<>(limit);
+        for (int i = 0; i < limit; i++) {
+            var item = frequentItems[i];
+            var row = new LinkedHashMap<>(2);
```

Comment:
Is a linked map here really needed?
```suggestion
            var row = new HashMap<>(2);
```

Reply:
👍